### ISSUE TITLE
Add a `load_all()` method

### ISSUE THREAD

**Issue by Erhannis** [collerek/ormar#107]
**Is your feature request related to a problem? Please describe.**
Sometimes I want the data of a given object and all its fields, recursively.  For example, sending the data off-server.  At the moment, AFAIK, in order to do this I need to manually load each (foreign) field of each class.

**Describe the solution you'd like**
I'd like a method, e.g. `load_all()`, which recursively calls `load()` and `all()` as needed on fields, and THEIR fields, and so forth, such that all of the data is present if I were to call e.g. `json()`.  (The solution need not actually call `load()` and `all()`, so long as the result is equivalent.)

There may be difficulties regarding circular references.  What I'd normally expect is for repeat references to be set to the same object instance previously loaded.  If that is currently infeasible - say, if the `__str__()` method and others would infinite loop, and it's a widespread, difficult problem - then I think it would be acceptable to use an object with only the primary key(s) loaded, when encountering a previously-loaded object.

**Describe alternatives you've considered**
I've considered explicitly, manually loading the fields of an object tree in relevant places.
I've also considered writing a function to try to do it for me.
The former seems like a lot of work, and brittle to boot.  The latter might work - it'd be a little tricky, though, and I'm told that there might be a more efficient way than how I'd be liable to implement it myself.  (My own implementation would consist primarily of repeated calls to `load()` and `all()`.)

**Additional context**
~~Add any other context or screenshots about the feature request here.~~

**Comment by collerek** [comment]
Added in 0.9.6 - please update and check

**Comment by Erhannis** [comment]
I did encounter one bug - when I upgraded, and used `.load_all(follow=True)` on my unmodified model, I got an error `ormar.exceptions.ModelError: Unknown field '[name of one of my fields]' for model Link`, where `Link` is presumably that dynamic linking class I made:

```
def link(tblname: str):
    class Link(ormar.Model):
        class Meta(BaseMeta):
            tablename = tblname
    return Link
```

I think maybe `through` doesn't respect `related_name`, because when I removed the `related_name` for the field to which the name given in the error most likely was intended to refer (i.e., when I removed the custom name that seemed to be broken), and updated the rest of my code accordingly, `.load_all(follow=True)` appeared to work fine.  So, that's good news, anyway, it seems to work aside from the `related_name` thing.  (Though I should probably do more than one single test, I suppose.)

Also - perhaps `load_all` could be present on `QuerysetProxy`, so we don't need to loop through the elements thereof and call `.load_all(follow=True)` on each?  Or, is there a recommended way of doing it I'm not familiar with?

**Comment by collerek** [comment]
You cannot set name of through field with `related_name`, if you refer to Link model in your example the trough name of the field will be `link`, if you mean `related_name` during creation of `ManyToMany` field it refers to name of the field that will be created on related model (the one passed as `to` parameter, reversed side). 

If you want to change the name of through field you need to simple change the name of `Through` class as the name is the `Through` class name lowercase. In your factory function you can do it by dynamically creating classes with given name.

When you `load_all` without excluding anything out it should include also the models to which queryset proxy refers. It won't load the same relation twice if you have loops in nested relations but apart from that it should go as deep as lack of circular references allows.

Or you mean you would like to `load_all()` only the relation with nested models without reloading the main model?

**Comment by Erhannis** [comment]
TL;DR - re: `related_name` etc., I changed my `link` code to make the names of the generated classes different from each other, and the problem went away.  Something still feels fishy, because that's not how I solved it in my real code, and I haven't observed any problems with name collisions _before_ the update - but the `link` change seems to fix it, so I'm not inclined to spend more time on the issue.

Original post, long version, probably unnecessary:

Yeah, I meant `related_name` of `ManyToMany`, sorry.  Keep in mind my code _works_, until I call `load_all` - or, at least, it used to work.  After more testing, I'm less sure of the conditions under which it now works.  So like:

```
def link(tblname: str):
    class Link(ormar.Model):
        class Meta(BaseMeta):
            tablename = tblname
    return Link

class CC(ormar.Model):
    class Meta(BaseMeta):
        tablename = "ccs"
    id: UUID = ormar.UUID(primary_key=True, default=uuid4)
    name: str = ormar.Text(default="")
    contents: str = ormar.Text(default="")

class CB(ormar.Model):
    class Meta(BaseMeta):
        tablename = "cbs"
    id: UUID = ormar.UUID(primary_key=True, default=uuid4)
    name: str = ormar.Text(default="")
    cCs: List[CC] = ormar.ManyToMany(CC, through=link("link_bc"), related_name="cBs")

class CA(ormar.Model):
    class Meta(BaseMeta):
        tablename = "cas"
    id: UUID = ormar.UUID(primary_key=True, default=uuid4)
    name: str = ormar.Text(default="")
    cBs: List[CB] = ormar.ManyToMany(CB, through=link("link_ab"), related_name="cAs")
```

```
In [2]: ca = await CA().save()

In [3]: cb = await CB().save()

In [4]: cc = await CC().save()

In [5]: await ca.cBs.add(cb)

In [6]: await cb.cCs.add(cc)

In [7]: await ca2.load_all(follow=True)
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py in __init__(self, *args, **kwargs)
    142         try:
--> 143             new_kwargs: Dict[str, Any] = {
    144                 k: self._convert_json(

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py in <dictcomp>(.0)
    145                     k,
--> 146                     self.Meta.model_fields[k].expand_relationship(
    147                         v, self, to_register=False,

KeyError: 'ca'

During handling of the above exception, another exception occurred:

ModelError                                Traceback (most recent call last)
<ipython-input-12-fe88e42d38c1> in <module>
----> 1 await ca2.load_all(follow=True)

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\model.py in load_all(self, follow, exclude, order_by)
    319         instance = await queryset.select_related(relations).get(pk=self.pk)
    320         self._orm.clear()
--> 321         self.update_from_dict(instance.dict())
    322         return self

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py in update_from_dict(self, value_dict)
    661         """
    662         for key, value in value_dict.items():
--> 663             setattr(self, key, value)
    664         return self
    665

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py in __setattr__(self, name, value)
    211         elif name in object.__getattribute__(self, "_orm"):
    212             model = (
--> 213                 object.__getattribute__(self, "Meta")
    214                 .model_fields[name]
    215                 .expand_relationship(value=value, child=self)

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in expand_relationship(cls, value, child, to_register)
    491         }
    492
--> 493         model = constructors.get(  # type: ignore
    494             value.__class__.__name__, cls._construct_model_from_pk
    495         )(value, child, to_register)

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in _extract_model_from_sequence(cls, value, child, to_register)
    341         :rtype: List["Model"]
    342         """
--> 343         return [
    344             cls.expand_relationship(  # type: ignore
    345                 value=val, child=child, to_register=to_register,

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in <listcomp>(.0)
    342         """
    343         return [
--> 344             cls.expand_relationship(  # type: ignore
    345                 value=val, child=child, to_register=to_register,
    346             )

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in expand_relationship(cls, value, child, to_register)
    491         }
    492
--> 493         model = constructors.get(  # type: ignore
    494             value.__class__.__name__, cls._construct_model_from_pk
    495         )(value, child, to_register)

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in _construct_model_from_dict(cls, value, child, to_register)
    393         if len(value.keys()) == 1 and list(value.keys())[0] == cls.to.Meta.pkname:
    394             value["__pk_only__"] = True
--> 395         model = cls.to(**value)
    396         if to_register:
    397             cls.register_relation(model=model, child=child)

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py in __init__(self, *args, **kwargs)
    141         # also remove property fields values from validation
    142         try:
--> 143             new_kwargs: Dict[str, Any] = {
    144                 k: self._convert_json(
    145                     k,

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py in <dictcomp>(.0)
    144                 k: self._convert_json(
    145                     k,
--> 146                     self.Meta.model_fields[k].expand_relationship(
    147                         v, self, to_register=False,
    148                     ),

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in expand_relationship(cls, value, child, to_register)
    491         }
    492
--> 493         model = constructors.get(  # type: ignore
    494             value.__class__.__name__, cls._construct_model_from_pk
    495         )(value, child, to_register)

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in _extract_model_from_sequence(cls, value, child, to_register)
    341         :rtype: List["Model"]
    342         """
--> 343         return [
    344             cls.expand_relationship(  # type: ignore
    345                 value=val, child=child, to_register=to_register,

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in <listcomp>(.0)
    342         """
    343         return [
--> 344             cls.expand_relationship(  # type: ignore
    345                 value=val, child=child, to_register=to_register,
    346             )

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in expand_relationship(cls, value, child, to_register)
    491         }
    492
--> 493         model = constructors.get(  # type: ignore
    494             value.__class__.__name__, cls._construct_model_from_pk
    495         )(value, child, to_register)

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in _construct_model_from_dict(cls, value, child, to_register)
    393         if len(value.keys()) == 1 and list(value.keys())[0] == cls.to.Meta.pkname:
    394             value["__pk_only__"] = True
--> 395         model = cls.to(**value)
    396         if to_register:
    397             cls.register_relation(model=model, child=child)

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py in __init__(self, *args, **kwargs)
    141         # also remove property fields values from validation
    142         try:
--> 143             new_kwargs: Dict[str, Any] = {
    144                 k: self._convert_json(
    145                     k,

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py in <dictcomp>(.0)
    144                 k: self._convert_json(
    145                     k,
--> 146                     self.Meta.model_fields[k].expand_relationship(
    147                         v, self, to_register=False,
    148                     ),

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in expand_relationship(cls, value, child, to_register)
    491         }
    492
--> 493         model = constructors.get(  # type: ignore
    494             value.__class__.__name__, cls._construct_model_from_pk
    495         )(value, child, to_register)

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\fields\foreign_key.py in _construct_model_from_dict(cls, value, child, to_register)
    393         if len(value.keys()) == 1 and list(value.keys())[0] == cls.to.Meta.pkname:
    394             value["__pk_only__"] = True
--> 395         model = cls.to(**value)
    396         if to_register:
    397             cls.register_relation(model=model, child=child)

c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py in __init__(self, *args, **kwargs)
    153             }
    154         except KeyError as e:
--> 155             raise ModelError(
    156                 f"Unknown field '{e.args[0]}' for model {self.get_name(lower=False)}"
    157             )

ModelError: Unknown field 'ca' for model Link
```

Now, I WAS going to say "If I remove the `related_name` fields, everything works", because I think I did that with my real code and I remember it working.  However, I did that with the test code, and it did NOT work; the same error remained.  If I instead remove my `through` parameter, THEN things work.  Is it related to your comment "Note also that i.e. `fastapi` don't work properly with multiple classes with same name from same file."?  I don't think I've had problems with the dynamic linking class, before, though....  Still, after some testing, the problem goes away if I update my `link` code to distinguish the class names:

```
def link(tblname: str):
    class Link(ormar.Model):
        class Meta(BaseMeta):
            tablename = tblname
    Link.__name__ += tblname
    Link.__qualname__ += tblname
    Link.Meta.__qualname__ = Link.__qualname__ + ".Meta"
    return Link
```

At this point I went back up and wrote the TL;DR.

**Comment by collerek** [comment]
As for link tables this is related with filed registration - `Through` field is also registered on both sides -> and since your `CB` model has links to both `CA` and `CC` the link field is overwritten (same name for links CA-CB and CB-CC) hence the error. That should be an exception with `ModelDefinitionError` -> will add this, thanks for catching that! :)

Note also that after update `Through` model is optional (if you don't have any additional fields there) so you can skip the generation and `ormar` does it for you, default names of the fields are both classes names lowercase (so in your example link fields would be `cacb` and `cbcc` respectively (and table names would be `cas_cbs` and `cbs_ccs`, classes names `CACB` and `CBCC`, default names are in docs)).

Ahh, now I get this -> you don't want to reload given instance but load all relations during query, yeah it should be easy to implement since I implemented `load_all()` in a very similar way (it uses the `QuerySet` under the hood), can you please open enhancement issue for `load_all` from `QuerySet`?

**Comment by Erhannis** [comment]
Oh, ok; good to know.  Hopefully that accounts for the behavior I was seeing earlier; at this point I can't tell anymore, haha.

And good to know; default `through` models seem convenient, thanks.

Yeah, I'll open the enhancement; thanks.
